keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -514,310 +481,126 @@ spec:
+To avoid any future conflicts, we reserve any attributes with the ".k8s.io" domain suffix
+for future use and standardization by Kubernetes. This could be used to describe
 topology across resources from different vendors, for example, but this is out-
 of-scope for now.
 
-**Note:** If a driver needs to reduce resource capacity, then there is a risk
-that a claim gets allocated using that capacity while the kubelet is updating a
-`ResourceSlice`. The implementations of structured models must handle
-scenarios where more resources are allocated than available. The kubelet plugin
-of a DRA driver ***must*** double-check that the allocated resources are still
+**Note:** If a driver needs to remove a device or change its attributes,
+then there is a risk that a claim gets allocated based on the old
+`ResourceSlice`. The scheduler must handle
+scenarios where more devices are allocated than available. The kubelet plugin
+of a DRA driver ***must*** double-check that the allocated devices are still
 available when NodePrepareResource is called. If not, the pod cannot start until
-the resource comes back. Treating this as a fatal error during pod admission
+the device comes back. Treating this as a fatal error during pod admission
 would allow us to delete the pod and trying again with a new one.
 
 ### Using structured parameters
 
-The following is an example CRD which the developer of the
-`cards.dra.example.com` DRA driver might define as a valid claim parameters
-object for requesting access to its GPUs:
+A ResourceClaim is a request to allocate one or more devices. Each request in a
+claim may reference a pre-defined DeviceClass to narrow down which devices are
+desired or describe that directly in the request itself through a device
+selector. A device selector is a CEL expression that must evaluate to true if a
+device satisfies the request. A special `device` variable provides access to
+the attributes of a device.
 
-```yaml
-kind: CardParameters
-apiVersion: dra.example.com/v1alpha1
-metadata:
-  name: my-parameters
-  namespace: user-namespace
-  uid: foobar-uid
-...
-spec:
-  minimumRuntimeVersion: v12.0.0
-  minimumMemory: 32Gi
-  # "sharing" is a configuration parameter that does not
-  # get translated into the selector below.
-  sharing:
-    strategy: TimeSliced
-```
-
-Note that all fields in this CRD can be fully validated since it is owned by
-the DRA driver itself. This includes value ranges that are specific to the
-underlying hardware. There's no risk of using invalid attribute names because
-only the fields shown here are valid.
-
-With this CRD in place, a DRA driver controller is able to convert instances of
-it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
-is able to understand.
+To correlate different devices, a claim may have "match attributes". Those are
+the names of attributes whose values must be the same for all devices that get
+allocated for the claim.
```

Comment:
What if I only want to correlate attributes for a subset of the devices in the claim?

Specifically, I'm thinking of the following (real-world) use-case, where I want 2 MIG devices from the same parent GPU and each of them wants a VF from the NIC on the same PCIe root complex:
```
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: mig-devices-and-nics
spec:
  requests:
  - name: mig-0
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"
    
  - name: mig-1
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"

  - name: nic-0
    deviceClassName: rdma.nvidia.com
    deviceSelectors:
    - "device['sriovType'] == 'vf'"
    
  - name: nic-1
    deviceClassName: rdma.nvidia.com
    deviceSelectors:
    - "device['sriovType'] == 'vf'"

  requirements:
  - match:
      attribute: parentUUID # For the MIG devices
  - match:
      attribute: k8s.io/pcieRoot # For all devices
```

Reply:
We cannot do this. Ways to accomplish this:

1) Allow cross-claim match attributes, in addition to within-claim-across-requests. Then, you would create two claims, one for the MIGs aligned on parentUUID, and one for the NICs, and you would specify a match across the claims of the pcieRoot. This requires podspec changes to allow match attributes.
2) Treat missing attributes in the match list as empty, and fully-qualify (with driver name) the parentUUID attribute with the `gpu.nvidia.com`. Since empty == empty, those would always match, and so these attribute matches would apply across all requests.
3) Match attributes becomes a "match expression" so the CEL can include some logic to get the result you want
4) Some sort of claim nesting...

For simplicity, 2) seems like the right call to me.
To do it, we would need: 1) to allocate the, because it would require either  *cross-claim* match attributes. Which we have not done because it requires putting match attributes into the pod spec.

Reply:
Hmm...what if the claim author could create "variables" which were CEL expressions that could then be used anywhere attributes can be used. That would solve it, too, while keeping the simplicity of the "match on list of attributes" API for most cases. Although maybe since an attribute name is basically a simple CEL expression, 3) is OK as it is.

Reply:
I think we want a "match expression" for more advanced cases. We talked about that before, we just haven't figured out what the CEL expression would look like. "Match attributes" are the simple solution that we can implement now. They have value also when we later add a "match expression" because they easier to use.

Reply:
> What if I only want to correlate attributes for a subset of the devices in the claim?

Your proposal from https://github.com/kubernetes/enhancements/pull/4709#discussion_r1632269856 could cover that.

Reply:
Yeah, except taht I think `match.attribute` is now flattened to `matchAttribute`, so:
```yaml
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: mig-devices-and-nics
spec:
  requests:
  - name: mig-0
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"
    
  - name: mig-1
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"

  - name: nic-0
    deviceClassName: rdma.nvidia.com
    deviceSelector: "device['sriovType'] == 'vf'"
    
  - name: nic-1
    deviceClassName: rdma.nvidia.com
    deviceSelector: "device['sriovType'] == 'vf'"

  constraints:
  - requests: ["mig-0", "mig-1"]
    matchAttribute: parentUUID 
  - matchAttribute: k8s.io/pcieRoot 
```

Reply:
We use "RequestName" elsewhere, so I am using "RequestNames" (not "Requests") also here.

Reply:
Actuall, thinking about this a bit more -- should it be `DeviceName` instead of `RequestName` since the thing we are requesting is always a `device` now? Likewise, should `requests` become `devices` or am I taking this too far...

Reply:
it also helps justify a (possible) decision to drop `device.` from the CEL expressions

Reply:
I find it clearer to say "a claim requests devices" and then actually have a "requests" field. But I can also live with renaming to "devices".

The API convention is `<kind>Name` when referencing some separate object with a known kind, so we shouldn't rename `deviceClassName` -> `className`.

Reply:
I think it makes sense to rename to `devices` because:
- If we have some `gizmos` in the future, we may want different request structures
- When (if) we add per-device allocatable resources, we probably want to use `requests` to represent those. Which means we would have `claim.spec.requests[*].requests` which is kind of weird. `claim.spec.devices[*].requests` will make more sense.
- For constraints and configs, we will then have `claim.spec.constraints[*].deviceNames` - when we have `gizmos` we may want to apply constraints differently

Reply:
Suppose we rename "requests" to "devices". What do we call the entries in that slice? A "device"?

More specifically, how do we refer to their name? @johnbelamaric  suggested "deviceNames" for constraints and configs and presumably "deviceName" for  `RequestAllocationResult` and the core `ResourceClaim` for a container.

That's a bit problematic:
- The `RequestAllocationResult` already contains a `deviceName` field, so there is a name clash right there in the same struct. Besides, the KEP becomes harder to understand when "device name" means different things depending on the context.
- `deviceName` looks odd to me when multiple devices get allocated. Example:

```yaml
containers:
   ...
   claims:
   - name: my-gpus
     deviceName: gpus # more than one GPU device
```

Currently, that is:
```yaml
containers:
   ...
   claims:
   - name: my-gpus
     requestName: gpus
```

Perhaps `deviceRequests` instead of just `requests`? Then we can use `deviceRequestName[s]` and call the things in the slice "device requests".

Reply:
Yeah, good points. I also didn't love `deviceName` referring to multiple devices, and we have already used it elsewhere. `deviceRequests` is sort of long. :shrug: 

I am ok with either leaving it `requests` or changing it to `deviceRequests`. If forced to pick, I would probably pick `deviceRequests` for the reasons I gave earlier.

Reply:
I'm okay with `deviceRequests`, no strong opinion.

@klueska?

Reply:
Okay. I'll keep the thread open as a reminder and for other reviewers.

Reply:
I'm not a fan of the nesting approach myself. Mostly because it is less flexible (what happens when I want to match an attribute from one element of in group 1 and one element in group 2)? But also because I find it confusing to have to specify constraints at two different levels. In my mental model, I first set up the requests to ask for a set of devices, then I come back and specify the constraints that need to be met between subsets of them.

Reply:
Yeah, I see your point Tim, but I actually think the nesting is not any easier to read or understand, and it is less flexible.

Kevin's mental model is similar to mine. It's "filter down to the broad set of devices I want", then "align that set as needed".

Reply:
> what happens when I want to match an attribute from one element in group 1 and one element in group 2

I'd say you laid out your groups wrong.  Or I'd say "show me the use case".

> I actually think the nesting is not any easier to read or understand

OK, well if that's the consensus, then my point is moot.  *I* find it easier to read, but *I* dreamed it up, so ...

I do still want to argue naming and structure for expressions.  Copying from Kevin's comment above:

```
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: mig-devices-and-nics
spec:
  requests:
  - name: mig-0
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"
    
  - name: mig-1
    deviceClassName: gpu.nvidia.com
    deviceSelectors: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"

  - name: nic-0
    deviceClassName: rdma.nvidia.com
    deviceSelector: "device['sriovType'] == 'vf'"
    
  - name: nic-1
    deviceClassName: rdma.nvidia.com
    deviceSelector: "device['sriovType'] == 'vf'"

  constraints:
  - requests: ["mig-0", "mig-1"]
    matchAttribute: parentUUID 
  - matchAttribute: k8s.io/pcieRoot 
```

could be:

```
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: mig-devices-and-nics
spec:
  requests:
  - name: mig-0
    deviceClassName: gpu.nvidia.com
    constraints:
      expression: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"
    
  - name: mig-1
    deviceClassName: gpu.nvidia.com
    constraints:
      expression: "device['type'] == 'MIG' && device['profile'] == '3g.20gb'"

  - name: nic-0
    deviceClassName: rdma.nvidia.com
    constraints:
      expression: "device['sriovType'] == 'vf'"
    
  - name: nic-1
    deviceClassName: rdma.nvidia.com
    constraints:
      expression: "device['sriovType'] == 'vf'"

  constraints:
  - requests: ["mig-0", "mig-1"]
    matchAttribute: parentUUID 
  - matchAttribute: k8s.io/pcieRoot 
```

Reply:
Either that, or they should be the same object. So, I allow `expression` and `matchAttribute` in both places. But `requests` wouldn't make sense in the per-request constraints. And you could end up with unsatisfiable combinations (though you can already write and always false CEL expression; but that's within an expression, not across two different fields).